File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Float64Vec.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Float64Vec {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class Float64VecCleaner implements Runnable {

        MemorySegment segment;
        Float64VecCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Float64Vec_destroy(this.segment);
        }
    }

    Float64Vec() {}
    Float64Vec(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static Float64Vec new_(double[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_DOUBLE, v);
            var vView = DiplomatF64View.allocate(Arena.ofAuto());
            DiplomatF64View.len(vView, vLen);
            DiplomatF64View.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newBool(boolean[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            byte[] vByteArray = new byte[vLen];
            for (int i = 0; i < vLen; i++) {
                vByteArray[i] = (byte) (v[i] ? 1 : 0);
            }
            var vData = Arena.ofAuto().allocateFrom(JAVA_BYTE, vByteArray);
            var vView = DiplomatBoolView.allocate(Arena.ofAuto());
            DiplomatBoolView.len(vView, vLen);
            DiplomatBoolView.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_bool(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newI16(short[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_SHORT, v);
            var vView = DiplomatI16View.allocate(Arena.ofAuto());
            DiplomatI16View.len(vView, vLen);
            DiplomatI16View.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_i16(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newU16(short[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_SHORT, v);
            var vView = DiplomatU16View.allocate(Arena.ofAuto());
            DiplomatU16View.len(vView, vLen);
            DiplomatU16View.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_u16(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newIsize(long[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_LONG, v);
            var vView = DiplomatIsizeView.allocate(Arena.ofAuto());
            DiplomatIsizeView.len(vView, vLen);
            DiplomatIsizeView.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_isize(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newUsize(long[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_LONG, v);
            var vView = DiplomatUsizeView.allocate(Arena.ofAuto());
            DiplomatUsizeView.len(vView, vLen);
            DiplomatUsizeView.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_usize(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static Float64Vec newF64BeBytes(byte[] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_BYTE, v);
            var vView = DiplomatU8View.allocate(Arena.ofAuto());
            DiplomatU8View.len(vView, vLen);
            DiplomatU8View.data(vView, vData);
    
            var nativeVal = somelib_h.Float64Vec_new_f64_be_bytes(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Float64Vec(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    
    public double[] asSlice() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.Float64Vec_as_slice(arena, internal);
            
            var data = dev.diplomattest.somelib.ntv.DiplomatF64View.data(nativeVal);
            var len = dev.diplomattest.somelib.ntv.DiplomatF64View.len(nativeVal);
            var returnVal = SliceUtils.doubleSliceToArray(nativeVal);
            return returnVal;
                    
        }
    }
    
    public void fillSlice(double[] v) {
        
        try (var arena = Arena.ofConfined()) {
            
            var vLen = v.length;
            var vData= Arena.ofAuto().allocateFrom(JAVA_DOUBLE, v);
            var vView = DiplomatF64View.allocate(Arena.ofAuto());
            DiplomatF64View.len(vView, vLen);
            DiplomatF64View.data(vView, vData);
    
            somelib_h.Float64Vec_fill_slice(internal, vView);
            
        }
    }
    
    public void setValue(double[] newSlice) {
        
        try (var arena = Arena.ofConfined()) {
            
            var newSliceLen = newSlice.length;
            var newSliceData= Arena.ofAuto().allocateFrom(JAVA_DOUBLE, newSlice);
            var newSliceView = DiplomatF64View.allocate(Arena.ofAuto());
            DiplomatF64View.len(newSliceView, newSliceLen);
            DiplomatF64View.data(newSliceView, newSliceData);
    
            somelib_h.Float64Vec_set_value(internal, newSliceView);
            
        }
    }
    
    public String toString_() {
        
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.Float64Vec_to_string(internal, writeable);
        var string = SliceUtils. readUtf8FromWriteable(writeable);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
    public double[] borrow() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.Float64Vec_borrow(arena, internal);
            
            var data = dev.diplomattest.somelib.ntv.DiplomatF64View.data(nativeVal);
            var len = dev.diplomattest.somelib.ntv.DiplomatF64View.len(nativeVal);
            var returnVal = SliceUtils.doubleSliceToArray(nativeVal);
            return returnVal;
                    
        }
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/Float64Vec.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface Float64VecLib: Library {
    fun Float64Vec_destroy(handle: Pointer)
    fun Float64Vec_new_bool(v: Slice): Pointer
    fun Float64Vec_new_i16(v: Slice): Pointer
    fun Float64Vec_new_u16(v: Slice): Pointer
    fun Float64Vec_new_isize(v: Slice): Pointer
    fun Float64Vec_new_usize(v: Slice): Pointer
    fun Float64Vec_new_f64_be_bytes(v: Slice): Pointer
    fun Float64Vec_new_from_owned(v: Slice): Pointer
    fun Float64Vec_as_slice(handle: Pointer): Slice
    fun Float64Vec_fill_slice(handle: Pointer, v: Slice): Unit
    fun Float64Vec_set_value(handle: Pointer, newSlice: Slice): Unit
    fun Float64Vec_to_string(handle: Pointer, write: Pointer): Unit
    fun Float64Vec_borrow(handle: Pointer): Slice
    fun Float64Vec_get(handle: Pointer, i: FFISizet): OptionDouble
}

class Float64Vec internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class Float64VecCleaner(val handle: Pointer, val lib: Float64VecLib) : Runnable {
        override fun run() {
            lib.Float64Vec_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<Float64VecLib> = Float64VecLib::class.java
        internal val lib: Float64VecLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun newBool(v: BooleanArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_bool(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newI16(v: ShortArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_i16(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newU16(v: UShortArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_u16(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newIsize(v: LongArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_isize(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newUsize(v: ULongArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_usize(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newF64BeBytes(v: ByteArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.borrow(v)
            
            val returnVal = lib.Float64Vec_new_f64_be_bytes(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newFromOwned(v: DoubleArray): Float64Vec {
            val vSliceMemory = PrimitiveArrayTools.move(v)
            
            val returnVal = lib.Float64Vec_new_from_owned(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Float64Vec(handle, selfEdges)
            CLEANER.register(returnOpaque, Float64Vec.Float64VecCleaner(handle, Float64Vec.lib));
            return returnOpaque
        }
    }
    
    fun asSlice(): DoubleArray {
        
        val returnVal = lib.Float64Vec_as_slice(handle);
            return PrimitiveArrayTools.getDoubleArray(returnVal)
    }
    
    fun fillSlice(v: DoubleArray): Unit {
        val vSliceMemory = PrimitiveArrayTools.borrow(v)
        
        val returnVal = lib.Float64Vec_fill_slice(handle, vSliceMemory.slice);
        
    }
    
    fun setValue(newSlice: DoubleArray): Unit {
        val newSliceSliceMemory = PrimitiveArrayTools.borrow(newSlice)
        
        val returnVal = lib.Float64Vec_set_value(handle, newSliceSliceMemory.slice);
        
    }
    
    override fun toString(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Float64Vec_to_string(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    fun borrow(): DoubleArray {
        
        val returnVal = lib.Float64Vec_borrow(handle);
            return PrimitiveArrayTools.getDoubleArray(returnVal)
    }
    
    internal fun getInternal(i: ULong): Double? {
        
        val returnVal = lib.Float64Vec_get(handle, FFISizet(i));
        return returnVal.option()
    }

    operator fun get(index: ULong): Double {
        val returnVal = getInternal(index)
        if (returnVal == null) {
            throw IndexOutOfBoundsException("Index $index is out of bounds.")
        } else {
            return returnVal
        }
    }

}